### PR TITLE
Integration of the multithreading functionality of the IoT Agent Node Lib into the IoT Agent - JSON

### PR DESCRIPTION
**Pull request by flopezag**
#### REVIEWER
@fgalan 

#### DESCRIPTION
- Modify the IoTAgent-json to deal with the option multi-thread.
- Correct bug about accessing configuration data of MQTT with not defined MQTT.
- Add the corresponding new variable in the config.js (multiCore) with description.
- Change in the Readme.md (Docker) and Installation Guide regarding High Performance execution of the IoT Agent - JSON.
- Create 2 examples files of the user limits and sysctl configuration for high performance execution.

#### TESTING
Unit tests, E2E tests in single-thread (legacy options) and multi-thread:
- Obtain IoT Agent JSON version info
- Provisioning Service
- Retrieve Service
- Provisioning Device
- Retrieve all devices
- Retrieve 1 device
- Send Measurement (*5)
- (Orion) Get list of entity types
- (Orion) Get list of entities
- (Orion) Get measurement
- Remove devide (*3)
- Retrieve all devices
- Remove service (*3)
- Retrieve services
- Obtain IoT Agent JSON version info

### CODE REVIEW

File: docker/README.md [hunk 1]
```diff
@@ -183,3 +183,43 @@ Currently, this the `_FILE` suffix is supported for:
...
+> Note: The --ulimit flags only work on Docker 1.6 or later.
+
+Nevertheless, you have to "request" more resources to Kubernetes (i.e. multiple cores), which might be more difficult
+for [Kubernetes](https://kubernetes.io/) to schedule than a few different containers requesting one core (or less...)
+each (which it can, in turn, schedule on multiple nodes, and not necessarily look for one node with enough available
+cores).
+
+If you want to get more details about the configuration of the system and node.js for high performance scenarios, please
+refer to the
+[Installation Guide](https://iotagent-node-lib.readthedocs.io/en/latest/installationguide/index.html#configuration).
```
Comment (fgalan) [inline docker/README.md]: I understand this maps to:

https://github.com/telefonicaid/iotagent-node-lib/blob/e7ef0e3ce1544e3940bd3c50b8b4a35beae9204e/doc/installationguide.md#configuration

However, it doesn't currently include information about "high performance scenarios"...

Is this paragraph correct? Do you plan to add such information in installationguide.md#configuration in a PR not yet caste (in which case this should be marked in this paragrpah as "work in progress")

File: docs/installationguide.md [hunk 2]
```diff
@@ -98,7 +98,9 @@ There are also some global configuration options:
 -   **configRetrieval**: this flag indicates whether the incoming notifications to the IoTAgent should be processed
     using the bidirectionality plugin from the latest versions of the library or the JSON-specific configuration
     retrieval mechanism (described in the User Manual). Simultaneous use of both mechanisms is not allowed.
--   **compressTimestamp**: this flags enables the timestamp compression mechanism, described in the User Manual.
+-   **compressTimestamp**: this flag enables the timestamp compression mechanism, described in the User Manual.
+-   **multiCore**: this (optional) flag enables the execution of the IoT Agent in multi-cores if it is `true` or in
```
Comment (fgalan) [inline docs/installationguide.md]: Not fully sure but I think to remember that the criteria is that library variables are described in library installation guide documentation. In agents repositories, only variables introduced by the agent (i..e mqtt and http things) are described.

@AlvaroVega do you remember how we are dealing with this in docs?

File: docker/README.md (hunk 1 above)
Comment (fgalan) [reply docker/README.md]: Ok... I see the "high availaibility scanerios" section in the installation.md file of this repository. Thus, probably it's just a matter of adjusting the URL to: https://fiware-iotagent-json.readthedocs.io/en/latest/installationguide/index.html

File: lib/iotagent-json.js [hunk 3]
```diff
@@ -137,7 +137,7 @@ function start(newConfig, callback) {
                 iotAgentLib.addQueryMiddleware(iotAgentLib.dataPlugins.compressTimestamp.query);
             }
 
-            if (config.getConfig().mqtt.thinkingThingsPlugin) {
+            if (config.getConfig().mqtt && config.getConfig().mqtt.thinkingThingsPlugin) {
```
Comment (fgalan) [inline lib/iotagent-json.js]: I understand this is not related to the topic of the PR, right?

(It's fine adding it, just asking confirmation of scope)

File: package.json [hunk 4]
```diff
@@ -61,7 +61,7 @@
     "body-parser": "1.18.3",
     "dateformat": "3.0.3",
     "express": "~4.16.4",
-    "iotagent-node-lib": "git://github.com/telefonicaid/iotagent-node-lib.git#master",
+    "iotagent-node-lib": "git://github.com/flopezag/iotagent-node-lib.git#feature/cluster-support",
```
Comment (fgalan) [inline package.json]: To be revered once flopezag:fetaure/cluster-support lands into iotagent-node-lib master (just a remark, for not forgetting)

File: docs/installationguide.md (hunk 2 above)
Comment (AlvaroVega) [reply docs/installationguide.md]: Do you mead this: https://github.com/telefonicaid/iotagent-node-lib/blob/master/doc/installationguide.md#configuration-using-environment-variables ?

Comment (fgalan) [reply docs/installationguide.md]: Yes.

My point is that if a given config parameter / env var is described in iotagent-node-lib documentation (because its processing belong to the library) that config parameters / env var is not described in agent's documentation. Is that correct?

File: lib/iotagent-json.js (hunk 3 above)
Comment (flopezag) [reply lib/iotagent-json.js]: Right, I discover this bug making some tests. Due to it is so tricky to resolve, I decided to include in this PR.

File: CHANGES_NEXT_RELEASE [hunk 5]
```diff
@@ -1,4 +1,5 @@
-ADD PM2_ENABLED flag to Docker
-Upgrade NodeJS version from 8.16.0 to 8.16.1 in Dockerfile due to security issues
-
-NOTE: in the "upgrade iotagent-node-lib from X to Y" include a mention about NGSIv2 forwarding support and issue #250
+- ADD PM2_ENABLED flag to Docker.
+- Upgrade NodeJS version from 8.16.0 to 8.16.1 in Dockerfile due to security issues.
+- Configure multithreading using cluster nodejs module (through multicore config.js parameter or IOTA_MULTI_CORE
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]:
```suggestion
- NOTE: in the "upgrade iotagent-node-lib from X to Y" include a mention to configure multithreading using cluster nodejs module (through multicore config.js parameter or IOTA_MULTI_CORE)
```

This is due to the actual change is not in the agents code. So, the changelog in this case is a bit different. Don't worry about that, we will deal with it at release time :)

File: docs/installationguide.md [hunk 6]
```diff
@@ -192,6 +194,150 @@ The ones relating specific JSON bindings are described in the following table.
 `IOTA_MQTT_CA`, `IOTA_MQTT_CERT`, `IOTA_MQTT_KEY` environment variables should provide the filename of the file whose
 contents will be used for the configuration attribute.
 
+#### High performance configuration
```
[the remark this answers is not in the document]
Comment (flopezag) [reply docs/installationguide.md]: Not sure, due to this section makes reference to the physical (or virtual machine) configuration to allow high performance execution, nothing related to the software configuration (even in the node lib).

[the remark this answers is not in the document]
Comment (fgalan) [reply docs/installationguide.md]: Ok. NTC

File: CHANGES_NEXT_RELEASE (hunk 5 above)
Comment (flopezag) [reply CHANGES_NEXT_RELEASE]: Fixed in e75739a

File: docs/installationguide.md (hunk 2 above)
Comment (fgalan) [reply docs/installationguide.md]: I have checked this with @AlvaroVega . It is confirmed my point, so please remove the multiCore setting from this documentation.

Comment (flopezag) [reply docs/installationguide.md]: Fixed in 7a2bf08

**Comment by fgalan** [comment]
Merged to pre-landing branch to prepare for CI e2e testing.
